san/eventstudy.py

```diff
@@ -0,0 +1,277 @@
+import numpy as np
+import pandas as pd
+import seaborn as sns
+import matplotlib.pyplot as pyplot
+from datetime import timedelta
+
+""" Event study to  evaluate events or signals.
+
+The main parameters the event study function accepts are a
+pandas dataframe containing the price data of the observed
+projects and the benchmark (data) and dataframe containing
+the events(ev_data)that contains the data of occurance in
+the index and the name of the project for every date.
+"""
+
+def get_close_price(data, sid, current_date, day_number):
+    #: If we're looking at day 0 just return the indexed date
+    if day_number == 0:
+        return data.loc[current_date][sid]      ##.ix
+    #: Find the close price day_number away from the current_date
+    else:
+        #: If the close price is too far ahead, just get the last available
+        total_date_index_length = len(data.index)
+        #: Find the closest date to the target date
+        date_index = data.index.searchsorted(current_date + timedelta(day_number))
+        #: If the closest date is too far ahead, reset to the latest date possible
+        date_index = total_date_index_length - 1 if date_index >= total_date_index_length else date_index
+        #: Use the index to return a close price that matches
+        return data.iloc[date_index][sid]
+
+def get_first_price(data, starting_point, sid, date):
+    starting_day = date - timedelta(starting_point)
+    date_index = data.index.searchsorted(starting_day)
+    return data.iloc[date_index][sid]
+
+def remove_outliers(returns, num_std_devs):
+    return returns[~((returns-returns.mean()).abs()>num_std_devs*returns.std())]
+
+def get_returns(data, starting_point, sid, date, day_num):
+    #: Get stock prices
+    first_price = get_first_price(data, starting_point, sid, date)
+    close_price = get_close_price(data, sid, date, day_num)
+
+    #: Calculate returns
+    ret = (close_price - first_price)/(first_price + 0.0)
+    return ret
+
+########################## Betas
+
+def calc_beta(stock, benchmark, price_history):
+    """
+    Calculate our beta amounts for each security
+    """
+    stock_prices = price_history[stock].pct_change().dropna()
+    bench_prices = price_history[benchmark].pct_change().dropna()
+    aligned_prices = bench_prices.align(stock_prices,join='inner')
+    bench_prices = aligned_prices[0]
+    stock_prices = aligned_prices[1]
+    bench_prices = np.array( bench_prices.values )
+    stock_prices = np.array( stock_prices.values )
+    bench_prices = np.reshape(bench_prices,len(bench_prices))
+    stock_prices = np.reshape(stock_prices,len(stock_prices))
+    if len(stock_prices) == 0:
+        return None
+    m, b = np.polyfit(bench_prices, stock_prices, 1)
+    return m
+
+# Simple event study
+
+def eventstudy(data, ev_data,starting_point = 30, benchmark='bitcoin',
```

Comment:
that's one big function over 200+ lines - needs refactoring. Also let's use snake case for naming functions to be consistent